api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+// or excluded domains are to be constructed.
+type ExternalDNSDomain struct {
+	ExternalDNSDomainUnion `json:",inline"`
+
+	// FilterType marks the Name or Pattern field
+	// as an included or excluded set of domains.
+	//
+	// Note that excluded domains that
+	// have not been included will be
+	// ignored.
+	//
+	// This field accepts the following values:
+	//
+	//  "Include": Include the domain set specified
+	//  by name or pattern.
+	//
+	//  "Exclude": Exclude the domain set specified
+	//  by name or pattern.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	FilterType ExternalDNSFilterType `json:"filterType"`
+}
+
+// ExternalDNSDomainUnion describes optional fields of an External domain
+// that should be captured.
+// +union
+type ExternalDNSDomainUnion struct {
+	// MatchType specifies the type of match to be performed
+	// by ExternalDNS when determining whether or not to publish DNS
+	// records for a given source resource based on the resource's
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType"`
+
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
+	// match a set of domains.
+	//
+	// +optional
+	Pattern *string `json:"pattern,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Exact;Pattern
+type DomainMatchType string
+
+const (
+	DomainMatchTypeExact DomainMatchType = "Exact"
+	DomainMatchTypeRegex DomainMatchType = "Pattern"
+)
+
+// +kubebuilder:validation:Enum=Include;Exclude
+type ExternalDNSFilterType string
+
+const (
+	FilterTypeInclude ExternalDNSFilterType = "Include"
+	FilterTypeExclude ExternalDNSFilterType = "Exclude"
+)
+
+// ExternalDNSProvider specifies configuration
+// options for the desired ExternalDNS DNS provider.
+type ExternalDNSProvider struct {
+	ExternalDNSProviderUnion `json:",inline"`
+
+	// Credentials is a reference to a secret containing
+	// the relevant credentials for the given provider, to be
+	// used by an ExternalDNS controller for creating, updating,
+	// and viewing DNS records.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Credentials *corev1.LocalObjectReference `json:"credentials"`
+}
+
+// ExternalDNSProviderUnion describes optional fields for an ExternalDNS
+// provider that should be captured.
+// +union
+type ExternalDNSProviderUnion struct {
+	// Type describes which DNS provider
+	// ExternalDNS should publish records to.
+	// The following DNS providers are supported:
+	//
+	//  * AWS
```

Comment:
I'm surprised that you don't require structured information for each of these providers in order to make connections to them.  How do you find them?

Reply:
> don't require structured information for each of these providers
We don't at the moment, but that could change in the future. Is there something that immediately jumps out to you as missing? 
> How do you find them?
Could you elaborate on what you mean? 

cc @Miciah

Reply:
I've added structured information for each provider.